src/main/java/com/segment/analytics/android/integrations/optimizelyx/OptimizelyXIntegration.java

```diff
@@ -1,63 +1,226 @@
-////        Logger logger = analytics.logger(OPTIMIZELY_KEY);
-////
-////        OptimizelyIntegration integration =
-////            new OptimizelyIntegration(analytics, application, logger);
-////        if (listen) {
-////          Optimizely.addOptimizelyEventListener(integration);
-////        }
-////        return integration;
-////      }
-//
-////      @Override public String key() {
-////      return OPTIMIZELY_KEY;
-//////      }
-//    };
-//  }
-
-  public OptimizelyXIntegration(Analytics analytics, Context context, Logger logger) {
-    this.analytics = analytics;
-    this.context = context;
+  boolean isClientValid = false;
+  boolean trackKnownUsers;
+  static boolean nonInteraction;
+  boolean listen;
+  static final Options options = new Options().setIntegration(OPTIMIZELYX_KEY, false);
+  private Map<String, String> attributes = new HashMap<>();
+  private List<TrackPayload> trackEvents = new ArrayList<>();
+  private final Handler handler = new Handler();
+
+  public static Factory factory(OptimizelyManager manager) {
+    return new Factory(manager);
+  }
+
+  private static class Factory implements Integration.Factory {
+    private final OptimizelyManager manager;
+
+    Factory(OptimizelyManager manager) {
+      this.manager = manager;
+    }
+
+    @Override
+    public Integration<?> create(ValueMap settings, Analytics analytics) {
+      Logger logger = analytics.logger(OPTIMIZELYX_KEY);
+      return new OptimizelyXIntegration(analytics, manager, settings, logger);
+    }
+
+    @Override
+    public String key() {
+      return OPTIMIZELYX_KEY;
+    }
+  }
+
+  public OptimizelyXIntegration(
+      final Analytics analytics, OptimizelyManager manager, ValueMap settings, Logger logger) {
+    this.client = manager.getOptimizely();
     this.logger = logger;
+
+    trackKnownUsers = settings.getBoolean("trackKnownUsers", false);
+    nonInteraction = settings.getBoolean("nonInteraction", true);
+    listen = settings.getBoolean("listen", true);
+
+    if (client.isValid()) {
+      isClientValid = true;
+      if (listen) {
+        this.listener = createListener(analytics);
+        client.addNotificationListener(listener);
+      }
+      return;
+    }
+    if (!client.isValid()) {
+      pollOptimizelyClient(manager, analytics);
+    }
   }
 
-  @Override public void identify(IdentifyPayload identify) {
+  @Override
+  public void identify(IdentifyPayload identify) {
+    super.identify(identify);
+
+    if (!isNullOrEmpty(identify.traits())) {
+      attributes = identify.traits().toStringMap();
+    }
   }
 
-  @Override public void track(TrackPayload track) {
+  @Override
+  public void track(TrackPayload track) {
+    super.track(track);
+
+    int queueSize = trackEvents.size();
+
+    synchronized (this) {
+      if (!isClientValid) {
+        logger.verbose("Optimizely not initialized. Enqueueing action: %s", track);
+        if (queueSize < 100) {
+          trackEvents.add(track);
```

Comment:
The two blocks have shared code. Should refactor it to be DRY. Also should log if a message is being dropped.